File: alamatic/parser.py

from alamatic.scanner import Scanner
from alamatic.ast import *
import alamatic.diagnostics as diag


def make_p_expr_binary_op(name, operator_map, next_name, allow_chain=True):
    this_parser = None

    def this_parser(self):
        scanner = self.scanner
        full_range = scanner.begin_range()

        next_parser = getattr(self, next_name)

        lhs = next_parser()

        peek = scanner.peek()
        if peek[0] in ("IDENT", peek[1]) and peek[1] in operator_map:
            operator = peek[1]
            ast_class = operator_map[peek[1]]
        else:
            return lhs

        # Eat the operator token, since we already dealt with it above.
        scanner.read()

        # As a special case for the funky, two-token "is not" operator,
        # we try to eat up a "not" right after an "is".
        # (Yes, this is a bit of a hack. Oh well.)
        if operator == "is" and scanner.next_is_keyword("not"):
            operator = "is not"
            scanner.read()

        if allow_chain:
            rhs = this_parser(self)
        else:
            rhs = next_parser()

        return ast_class(full_range.end(), lhs, operator, rhs)

    this_parser.__name__ = name
    return this_parser


def make_p_expr_prefix_unary_op(name, operator_map, next_name):
    this_parser = None

    def this_parser(self):
        scanner = self.scanner
        full_range = scanner.begin_range()

        next_parser = getattr(self, next_name)

        peek = scanner.peek()
        if peek[0] in ("IDENT", peek[1]) and peek[1] in operator_map:
            operator = peek[1]
            ast_class = operator_map[peek[1]]
            scanner.read()
            # Try running ourselves again so that we can chain,
            # like "not not a". Isn't incredibly useful but it ought to
            # work anyway for consistency.
            operand = this_parser(self)
            return ast_class(full_range.end(), operand, operator)
        else:
            return next_parser()

    this_parser.__name__ = name
    return this_parser


class Parser(object):

    def __init__(self, scanner):
        self.scanner = scanner

    def parse_module(self):
        source_range, block, doc = self.p_toplevel()
        return Module(source_range, name, block, doc=doc)

    def parse_entry_file(self):
        source_range, block, doc = self.p_toplevel()
        return EntryFile(source_range, block, doc=doc)

    def parse_expression(self, allow_assign=False):
        try:
            expr = self.p_expression(allow_assign=allow_assign)
            scanner.require_eof()
        except diag.Diagnostic, err:
            expr = ErroredExpression(err)
        return expr

    def p_toplevel(self):
        scanner = self.scanner

        full_range = scanner.begin_range()

        # doc comments at the immediate start of a module are for the
        # module itself.
        doc = self.p_doc_comments()

        # To avoid ambiguity, a newline is required after a module docstring.
        ambiguous_module_docstring = False
        if doc is not None:
            if not (scanner.next_is_newline() or scanner.next_is_eof()):
                ambiguous_module_docstring = True

        stmts = self.p_statements(lambda s: s.next_is_eof())

        if ambiguous_module_docstring:
            # Treat the missing blank line as if it were an erroneous
            # statement at the beginning of the statement list.
            error = CompilerError(
                "Insert a blank link after the module doc comments, at ",
                pos_link(scanner.position()), ". If this was intended as a ",
                "doc comment for the following statement instead, add a ",
                "blank line before it."
            )
            error_stmt = ErroredStatement(error)
            stmts.insert(0, error_stmt)

        source_range = full_range.end()

        block = StatementBlock(stmts, source_range=source_range)
        return source_range, block, doc

    def p_doc_comments(self):
        scanner = self.scanner
        peek = scanner.peek()
        if peek[0] == "DOCCOMMENT":
            parts = []
            while peek[0] == "DOCCOMMENT":
                scanner.read()
                parts.append(peek[1])
                scanner.require_newline()
                peek = scanner.peek()
            return "\n".join(parts)
        else:
            # no doc comments at all is signalled as None, to differentiate
            # between this case an an *empty* doc comment.
            return None

    def p_statements(self, stop_test=lambda s: s.next_is_outdent()):
        scanner = self.scanner
        stmts = []
        while not stop_test(scanner):
            stmt = None
            try:
                stmt = self.p_statement()
            except diag.Diagnostic, err:
                stmt = ErroredStatement(err)
                scanner.skip_statement()
            if stmt is not None:
                stmts.append(stmt)

        return stmts

    def p_indented_block(self):
        scanner = self.scanner
        full_range = scanner.begin_range()
        scanner.require_punct(":")
        scanner.require_newline()
        scanner.require_indent()
        stmts = self.p_statements()
        scanner.require_outdent()
        return StatementBlock(stmts, full_range.end())

    def p_expr_list(self, terminator_punct, allow_assign=False):
        scanner = self.scanner
        exprs = []
        if scanner.next_is_punct(terminator_punct):
            scanner.read()
        else:
            while True:
                exprs.append(
                    self.p_expression(allow_assign=allow_assign)
                )
                if scanner.next_is_punct(terminator_punct):
                    scanner.read()
                    break
                scanner.require_punct(",")
                if scanner.next_is_punct(terminator_punct):
                    scanner.read()
                    break

        return ExpressionList(exprs)

    def p_arguments(self, terminator_punct):
        scanner = self.scanner
        expr_list = self.p_expr_list(terminator_punct, allow_assign=True)

        args = []
        kwargs = {}

        for expr in expr_list.exprs:
            if isinstance(expr, AssignExpr):
                if isinstance(expr.lhs, SymbolNameExpr):
                    name = expr.lhs.name
                    kwargs[name] = expr.rhs
                else:
                    raise CompilerError(
                        "Invalid keyword argument at ",
                        pos_link(self.lhs.position),
                    )
            else:
                args.append(expr)

        return Arguments(args, kwargs)

    def p_statement(self):
        scanner = self.scanner
        full_range = scanner.begin_range()

        if scanner.next_is_newline():
            # empty statement
            scanner.read()
            return None

        peek = scanner.peek()
        if peek[0] == "IDENT":
            ident = peek[1]
            if ident == "pass":
                scanner.read()
                scanner.require_newline()
                return PassStmt(full_range.end())
            if ident == "break":
                scanner.read()
                scanner.require_newline()
                return BreakStmt(full_range.end())
            if ident == "continue":
                scanner.read()
                scanner.require_newline()
                return ContinueStmt(full_range.end())
            if ident == "return":
                scanner.read()
                expr = None
                if not scanner.next_is_newline():
                    expr = self.p_expression()
                scanner.require_newline()
                return ReturnStmt(full_range.end(), expr)
            if ident == "if":
                return self.p_if_stmt()
            if ident == "while":
                scanner.read()
                expr = self.p_expression()
                block = self.p_indented_block()
                return WhileStmt(full_range.end(), expr, block)
            if ident == "for":
                return self.p_for_stmt()
            if ident == "var" or ident == "const":
                decl = self.p_data_decl()
                expr = None
                if scanner.next_is_punct('='):
                    scanner.read()
                    expr = self.p_expression()
                scanner.require_newline()
                return DataDeclStmt(full_range.end(), decl, expr)
            if ident == "func":
                decl = self.p_func_decl()
                block = self.p_indented_block()
                return FuncDeclStmt(full_range.end(), decl, block)

        expr = self.p_expression(allow_assign=True)
        scanner.require_newline()
        # FIXME: Should fail if an expression statement has no
        # direct side-effects... that is, if it's not a call or an
        # assignment node.
        return ExpressionStmt(full_range.end(), expr)

    def p_if_stmt(self):
        scanner = self.scanner
        full_range = scanner.begin_range()

        clauses = []

        if_range = scanner.begin_range()
        scanner.require_keyword("if")
        if_expr = self.p_expression()
        if_block = self.p_indented_block()

        clauses.append(IfClause(if_range.end(), if_expr, if_block))

        while scanner.next_is_keyword("elif"):
            elif_range = scanner.begin_range()
            scanner.read()
            elif_expr = self.p_expression()
            elif_block = self.p_indented_block()
            clauses.append(IfClause(elif_range.end(), elif_expr, elif_block))

        if scanner.next_is_keyword("else"):
            else_range = scanner.begin_range()
            scanner.read()
            else_block = self.p_indented_block()
            clauses.append(ElseClause(else_range.end(), else_block))

        return IfStmt(full_range.end(), clauses)

    def p_for_stmt(self):
        scanner = self.scanner
        full_range = scanner.begin_range()

        scanner.require_keyword("for")
        if scanner.next_is_keyword("var") or scanner.next_is_keyword("const"):
            target = self.p_data_decl()
        else:
            target = self.p_expression()

        scanner.require_keyword("in")
        source_expr = self.p_expression()

        block = self.p_indented_block()

        return ForStmt(full_range.end(), target, source_expr, block)

    def p_expression(self, allow_assign=False):
        scanner = self.scanner
        peek = scanner.peek()

        if allow_assign:
            return self.p_expr_assign()
        else:
            # skip straight to logical or, which is the next operator in
            # precedence order
            return self.p_expr_logical_or()

    def p_expr_factor(self):
        scanner = self.scanner
        source_range = scanner.begin_range()

        ret_expr = None

        ident_literals = {
            "true": True,
            "false": False,
            "null": None,
        }

        if scanner.next_is_punct("("):
            scanner.read()
            ret_expr = self.p_expression()
            scanner.require_punct(")")
        else:
            peek = scanner.peek()

            if peek[0] == "NUMBER":
                ret_expr = self.p_expr_number()
            elif peek[0] == "STRINGLIT":
                ret_expr = self.p_expr_string()
            elif peek[0] == "IDENT":
                scanner.read()
                if peek[1] in ident_literals:
                    ret_expr = LiteralExpr(
                        source_range.end(),
                        ident_literals[peek[1]],
                    )
                else:
                    ret_expr = SymbolNameExpr(source_range.end(), peek[1])
            else:
                raise CompilerError(
                    "Expected expression but found ",
                    scanner.token_display_name(scanner.peek()),
                    " at ", pos_link(source_range.start),
                )

        # Now handle calls, subscripts and attribute access, if present.
        # These can be chained indefinitely.
        while True:
            if scanner.next_is_punct("("):
                scanner.read()
                args = self.p_arguments(")")
                ret_expr = CallExpr(
                    source_range.end(),
                    ret_expr,
                    args,
                )
            elif scanner.next_is_punct("["):
                scanner.read()
                args = self.p_expr_list("]")
                ret_expr = SubscriptExpr(
                    source_range.end(),
                    ret_expr,
                    args,
                )
            elif scanner.next_is_punct("."):
                scanner.read()
                if scanner.peek()[0] != "IDENT":
                    raise CompilerError(
                        "Expected attribute name but got ",
                        scanner.token_display_name(scanner.peek()),
                        " at ", pos_link(scanner.position()),
                    )
                attr_name = scanner.read()[1]
                ret_expr = AttributeExpr(
                    source_range.end(),
                    ret_expr,
                    attr_name,
                )
            else:
                break

        return ret_expr

    p_expr_bitwise_not = make_p_expr_prefix_unary_op(
        "p_expr_bitwise_not",
        {
            "~": BitwiseNotExpr,
        },
        "p_expr_factor",
    )

    p_expr_sign = make_p_expr_prefix_unary_op(
        "p_expr_sign",
        {
            "-": SignExpr,
            "+": SignExpr,
        },
        "p_expr_bitwise_not",
    )

    p_expr_multiply = make_p_expr_binary_op(
        "p_expr_multiply",
        {
            "*": MultiplyExpr,
            "/": MultiplyExpr,
            "%": MultiplyExpr,
        },
        "p_expr_sign",
    )

    p_expr_sum = make_p_expr_binary_op(
        "p_expr_sum",
        {
            "+": SumExpr,
            "-": SumExpr,
        },
        "p_expr_multiply",
    )

    p_expr_shift = make_p_expr_binary_op(
        "p_expr_shift",
        {
            "<<": ShiftExpr,
            ">>": ShiftExpr,
        },
        "p_expr_sum",
    )

    p_expr_bitwise_and = make_p_expr_binary_op(
        "p_expr_bitwise_and",
        {
            "&": BitwiseAndExpr,
        },
        "p_expr_shift",
    )

    p_expr_bitwise_or = make_p_expr_binary_op(
        "p_expr_bitwise_or",
        {
            "|": BitwiseOrExpr,
        },
        "p_expr_bitwise_and",
    )

    p_expr_comparison = make_p_expr_binary_op(
        "p_expr_comparison",
        {

            "is": ComparisonExpr,
            # A special case inside make_p_expr_binary_op allows
            # "is" to also implement "is not". (Hacky, yes.)

            "<": ComparisonExpr,
            "<=": ComparisonExpr,
            ">": ComparisonExpr,
            ">=": ComparisonExpr,
            "!=": ComparisonExpr,
            "==": ComparisonExpr,
        },
        "p_expr_bitwise_or",
    )

    p_expr_logical_not = make_p_expr_prefix_unary_op(
        "p_expr_logical_not",
        {
            "not": LogicalNotExpr,
        },
        "p_expr_comparison",
    )

    p_expr_logical_and = make_p_expr_binary_op(
        "p_expr_logical_and",
        {
            "and": LogicalAndExpr,
        },
        "p_expr_logical_not",
    )

    p_expr_logical_or = make_p_expr_binary_op(
        "p_expr_logical_or",
        {
            "or": LogicalOrExpr,
        },
        "p_expr_logical_and",
    )

    p_expr_assign = make_p_expr_binary_op(
        "p_expr_assign",
        {
            "=": AssignExpr,
            "+=": AssignExpr,
            "-=": AssignExpr,
            "*=": AssignExpr,
            "/=": AssignExpr,
            "|=": AssignExpr,
            "&=": AssignExpr,
        },
        "p_expr_logical_or",
        allow_chain=False,
    )

    def p_expr_number(self):
        scanner = self.scanner
        full_range = scanner.begin_range()

        token = scanner.read()
        if token[0] != "NUMBER":
            raise CompilerError(
                "Expected number but got ",
                scanner.token_display_name(token),
                " at ", pos_link(full_range.start),
            )

        num_str = token[1]

        imaginary = False
        significand = None
        exponent = 0

        if num_str.startswith("0x"):
            try:
                significand = int(num_str[2:], 16)
            except ValueError:
                raise CompilerError(
                    num_str[2:], " is not a valid hexadecimal number, "
                    " at ", pos_link(full_range.start),
                )
        elif num_str.startswith("0b"):
            try:
                significand = int(num_str[2:], 2)
            except ValueError:
                raise CompilerError(
                    num_str[2:], " is not a valid binary number, "
                    " at ", pos_link(full_range.start),
                )
        elif num_str.startswith("0") and len(num_str) > 1:
            try:
                value = int(num_str[1:], 8)
            except ValueError:
                raise CompilerError(
                    num_str[1:], " is not a valid octal number, "
                    " at ", pos_link(full_range.start),
                )
        else:
            # Try to parse it as a base-10 int, and if that fails
            # try to parse as a float. If neither work, fail hard.
            try:
                significand = int(num_str, 10)
            except ValueError:
                try:
                    whole, frac = num_str.split(".", 2)
                    exponent = -len(frac)
                    significand = whole + frac  # this is string concatenation
                    significand = int(significand, 10)
                except ValueError:
                    raise CompilerError(
                        num_str, " is not a valid number, "
                        " at ", pos_link(full_range.start),
                    )

            if scanner.next_is_keyword("j"):
                scanner.read()
                imaginary = True

        if significand is None:
            # Should never happen
            raise Exception(
                "Failed to produce a value for number token " + num_str
            )

        return NumericLiteralExpr(
            full_range.end(),
            significand,
            exponent,
            imaginary,
        )

    def p_data_decl(self):
        scanner = self.scanner
        full_range = scanner.begin_range()

        if scanner.next_is_keyword("var"):
            decl_type = VarDeclClause
        elif scanner.next_is_keyword("const"):
            decl_type = ConstDeclClause
        else:
            raise CompilerError(
                "Expected declaration but got ",
                scanner.token_display_name(scanner.peek()),
                " at ", pos_link(full_range.start),
            )

        scanner.read()

        if scanner.peek()[0] != "IDENT":
            raise CompilerError(
                "Expected declaration name but got ",
                scanner.token_display_name(scanner.peek()),
                " at ", pos_link(scanner.position()),
            )

        name = scanner.read()[1]

        return decl_type(full_range.end(), name)

    def p_func_decl(self):
        scanner = self.scanner
        full_range = scanner.begin_range()

        scanner.require_keyword("func")

        if scanner.peek()[0] != "IDENT":
            raise CompilerError(
                "Expected function name but got ",
                scanner.token_display_name(scanner.peek()),
                " at ", pos_link(scanner.position()),
            )

        name = scanner.read()[1]

        params = []
        scanner.require_punct("(")
        if scanner.next_is_punct(")"):
            scanner.read()
        else:
            while True:
                if scanner.peek()[0] != "IDENT":
                    raise CompilerError(
                        "Expected parameter name but got ",
                        scanner.token_display_name(scanner.peek()),
                        " at ", pos_link(scanner.position()),
                    )
                param_range = scanner.begin_range()
                param_const = False
                param_named = False
                param_collector = False
                param_default_expr = None
                param_type_constraint_expr = None

                if scanner.next_is_keyword("const"):
                    scanner.read()
                    param_const = True

                if scanner.next_is_keyword("named"):
                    scanner.read()
                    param_named = True

                if scanner.next_is_keyword("const"):
                    if param_named:
                        raise CompilerError(
                            "'const' qualifier must appear before 'named',",
                            " not at ", pos_link(scanner.position()),
                        )
                    else:
                        raise CompilerError(
                            "Parameter must not be called 'const' at ",
                            pos_link(scanner.position()),
                        )

                param_name = scanner.read()[1]

                if scanner.next_is_punct("="):
                    scanner.read()
                    param_default_expr = self.p_expression()

                if scanner.next_is_keyword("when"):
                    scanner.read()
                    param_type_constraint_expr = self.p_expression()

                if scanner.next_is_punct("..."):
                    scanner.read()
                    param_collector = True

                params.append(
                    ParamDeclClause(
                        param_range.end(),
                        param_name,
                        const_required=param_const,
                        named=param_named,
                        default_expr=param_default_expr,
                        collector=param_collector,
                        type_constraint_expr=param_type_constraint_expr,
                    )
                )

                if scanner.next_is_punct(")"):
                    scanner.read()
                    break
                scanner.require_punct(",")
                if scanner.next_is_punct(")"):
                    scanner.read()
                    break

        return FuncDeclClause(full_range.end(), name, params)

File: ast/parser.go
package ast

import (
	"github.com/alamatic/alamatic/diag"
	"github.com/alamatic/alamatic/tokenizer"
)

type parser struct {
	peeker *tokenizer.TokenPeeker
}

func ParseModule(c <-chan tokenizer.Token) *Module {
	peeker := tokenizer.NewPeeker(c)
	parser := &parser{peeker}
	block, doc := parser.ParseTopLevel()
	return &Module{
		Block:       block,
		SourceRange: block.SourceRange,
		Doc:         doc,
	}
}

func (p *parser) ParseTopLevel() (*StatementBlock, DocString) {
	peeker := p.peeker

	fullRange := peeker.RangeBuilder()

	doc := p.ParseDocComments()

	// To avoid ambiguity, an empty line is always required after a module
	// docstring, so it can be distinguished from a declaration docstring
	// for the file's first declaration.
	var ambiguousModuleDocstringError *diag.Diagnostic
	if doc != "" {
		next := peeker.Peek()
		if next.Kind != tokenizer.NewLine && next.Kind != tokenizer.EOF {
			type ambiguousModuleDocstringDetails struct {
				diag.Message `Empty line required after module docstring`
			}

			ambiguousModuleDocstringError = &diag.Diagnostic{
				Level:       diag.Error,
				Details:     &ambiguousModuleDocstringDetails{},
				SourceRange: next.SourceRange(),
			}
		}
	}

	stmts := p.ParseStatements(func(next *tokenizer.Token) bool {
		// Stop at the end of the file
		return next.Kind == tokenizer.EOF
	})

	if ambiguousModuleDocstringError != nil {
		// Treat the missing blank line as a diagnostic statement at the
		// beginning of the statement list.
		origStmts := stmts
		stmts = make([]Statement, 0, len(origStmts)+1)
		stmts = append(stmts, &DiagnosticStmt{
			Diagnostics: []*diag.Diagnostic{
				ambiguousModuleDocstringError,
			},
		})
		stmts = append(stmts, origStmts...)
	}

	sourceRange := fullRange()

	return &StatementBlock{
		Statements:  stmts,
		SourceRange: sourceRange,
	}, doc
}

func (p *parser) ParseDocComments() DocString {
	return DocString("")
}

func (p *parser) ParseStatements(stop func(*tokenizer.Token) bool) []Statement {
	peeker := p.peeker

	if stop == nil {
		stop = func(t *tokenizer.Token) bool {
			return t.Kind == tokenizer.Outdent
		}
	}

	stmts := make([]Statement, 0, 10)

	for !stop(peeker.Peek()) {
		if peeker.Peek().Kind == tokenizer.BadOutdent {
			badOutdent := peeker.Read()
			err := &diag.Diagnostic{
				Level:       diag.Error,
				Details:     &InconsistentIndentation{},
				SourceRange: badOutdent.SourceRange(),
			}
			stmts = append(stmts, &DiagnosticStmt{
				Diagnostics: []*diag.Diagnostic{
					err,
				},
			})
			// TODO: Try to seek past the current statement
			// so we can recover and parse the rest of the file.
			break
		}
		stmt := p.ParseStatement()
		if stmt != nil {
			stmts = append(stmts, stmt)
		}
	}

	return stmts
}

func (p *parser) ParseIndentedBlock() *StatementBlock {
	peeker := p.peeker
	fullRange := peeker.RangeBuilder()

	parseErr := func() *StatementBlock {
		tokenRange := peeker.Peek().SourceRange()
		return &StatementBlock{
			Statements: []Statement{
				&DiagnosticStmt{
					Diagnostics: []*diag.Diagnostic{
						{
							Level:       diag.Error,
							Details:     &IndentedBlockExpected{},
							SourceRange: tokenRange,
						},
					},
				},
			},
			SourceRange: tokenRange,
		}
	}

	next := peeker.Peek()
	if next.Kind != tokenizer.Punct || len(next.Bytes) != 1 || next.Bytes[0] != ':' {
		return parseErr()
	}

	next = peeker.Peek()
	if next.Kind != tokenizer.NewLine {
		return parseErr()
	}

	next = peeker.Peek()
	if next.Kind != tokenizer.Indent {
		return parseErr()
	}

	stmts := p.ParseStatements(func(next *tokenizer.Token) bool {
		return next.Kind == tokenizer.Outdent
	})

	// Gobble the remaining outdent
	peeker.Read()

	return &StatementBlock{
		Statements:  stmts,
		SourceRange: fullRange(),
	}
}

func (p *parser) ParseStatement() Statement {
	peeker := p.peeker

	fullRange := peeker.RangeBuilder()
	next := peeker.Peek()

	if next.Kind == tokenizer.NewLine {
		// Empty statement, for which we don't bother to create an AST node
		peeker.Read()
		return nil
	}

	if next.Kind == tokenizer.Ident {
		// Some legal identifiers are actually keywords when appearing
		// at the beginning of a statement.
		switch string(next.Bytes) {
		case "pass":
			peeker.Read()
			stmt := &PassStmt{
				SourceRange: fullRange(),
			}
			return p.requireStatementEOL(stmt)
		case "break":
			peeker.Read()
			stmt := &BreakStmt{
				SourceRange: fullRange(),
			}
			return p.requireStatementEOL(stmt)
		case "continue":
			peeker.Read()
			stmt := &ContinueStmt{
				SourceRange: fullRange(),
			}
			return p.requireStatementEOL(stmt)
		case "return":
			return p.ParseReturnStmt()
		case "if":
			return p.ParseIfStmt()
		case "while":
			return p.ParseWhileStmt()
		case "for":
			return p.ParseForStmt()
		case "var", "const":
			return p.ParseDataDeclStmt()
		case "func", "proc":
			return p.ParseFuncDecl()
		}
	}

	expr := p.ParseExpression(true)
	stmt := &ExprStmt{
		Expression: expr,
	}
	// FIXME: Generate a diagnostic node if an expression statement
	// has no direct side-effects... that is, if it's not a call
	// or an assignment.
	return p.requireStatementEOL(stmt)
}

func (p *parser) ParseReturnStmt() Statement {
	peeker := p.peeker

	fullRange := peeker.RangeBuilder()

	// Assumed to be called only when the next token is "return"
	peeker.Read()
	var expr Expression
	next := peeker.Peek()
	if next.Kind != tokenizer.NewLine {
		expr = p.ParseExpression(false)
	}
	stmt := &ReturnStmt{
		Expression:  expr,
		SourceRange: fullRange(),
	}
	return p.requireStatementEOL(stmt)
}

func (p *parser) ParseIfStmt() Statement {
	peeker := p.peeker

	fullRange := peeker.RangeBuilder()

	// Assumed to be called only when the next token is "if"
	peeker.Read()

	clauses := make([]*IfClause, 0, 1)

	ifExpr := p.ParseExpression(false)
	ifBlock := p.ParseIndentedBlock()

	clauses = append(clauses, &IfClause{
		CondExpr: ifExpr,
		Block:    ifBlock,
	})

	for {
		next := peeker.Peek()
		if next.Kind != tokenizer.Ident || string(next.Bytes) != "elif" {
			break
		}

		peeker.Read()
		elseIfExpr := p.ParseExpression(false)
		elseIfBlock := p.ParseIndentedBlock()

		clauses = append(clauses, &IfClause{
			CondExpr: elseIfExpr,
			Block:    elseIfBlock,
		})
	}

	if next := peeker.Peek(); next.Kind == tokenizer.Ident && string(next.Bytes) == "else" {
		peeker.Read()
		elseBlock := p.ParseIndentedBlock()

		clauses = append(clauses, &IfClause{
			Block: elseBlock,
		})
	}

	stmt := &IfStmt{
		Clauses:     clauses,
		SourceRange: fullRange(),
	}
	return stmt
}

func (p *parser) ParseWhileStmt() Statement {
	panic("while not implemented")
}

func (p *parser) ParseForStmt() Statement {
	panic("for not implemented")
}

func (p *parser) ParseDataDeclStmt() Statement {
	panic("data decl not implemented")
}

func (p *parser) ParseFuncDecl() Statement {
	panic("func decl not implemented")
}

func (p *parser) ParseExpression(allowAssign bool) Expression {
	panic("expressions not yet implemented")
}

func (p *parser) requireStatementEOL(s Statement) Statement {
	next := p.peeker.Peek()
	if next.Kind != tokenizer.NewLine {
		// TODO: Try to seek to the beginning of the next statement
		// so we can attempt to parse the rest of the file.
		return &DiagnosticStmt{
			Diagnostics: []*diag.Diagnostic{
				&diag.Diagnostic{
					Level:       diag.Error,
					Details:     &NewlineExpected{},
					SourceRange: next.SourceRange(),
				},
			},
		}
	}

	return s
}
